javaslang/src/main/java/javaslang/Value.java

```diff
@@ -660,13 +668,46 @@ default CharSeq toCharSeq() {
      */
     default <K, V> Map<K, V> toMap(Function<? super T, ? extends Tuple2<? extends K, ? extends V>> f) {
         Objects.requireNonNull(f, "f is null");
-        if (isEmpty()) {
-            return HashMap.empty();
-        } else if (isSingleValued()) {
-            return HashMap.of(f.apply(get()));
-        } else {
-            return HashMap.ofEntries(Iterator.ofAll(this).map(f));
-        }
+		return ValueModule.toMap(this, HashMap.empty(), HashMap::of, HashMap::ofEntries, f);
```

Comment:
Very nice, thank you!

I think we should additionally check in all toMap and toSet methods that `this` Value is not already of type Map or Set, e.g.

```java
default <K, V> Map<K, V> toMap(Function<? super T, ? extends Tuple2<? extends K, ? extends V>> f) {
    Objects.requireNonNull(f, "f is null");
    if (Map.class.isAssignableFrom(getClass())) {
        return (Map<K, V>) this;
    } else {
        return ValueModule.toMap(this, HashMap.empty(), HashMap::of, HashMap::ofEntries, f);
    }
}
```

Maybe we can move the check into ValueModule.toMap / toSet by adding a mapType / setType parameter of type class? But generics in `Class<...>` could make problems. It could be just `Class<?>` and be cast to `M`...